Fix broken pkgIndex.tcl on Windows and fix macOS builds for clang>=16

conda-forge-admin opened the pull request:
In https://github.com/conda-forge/r-base-feedstock/pull/278#discussion_r1649598776 we have noticed an issue with loading Tktable due to a malformed `pkgIndex.tcl`.

This is easily resolved:
> Tktable 2.8 had VERSION defined via generic/version.h which got then
> reused in win/makefile.vc.
> Tktable 2.10 has it renamed to PACKAGE_VERSION but win/makefile.vc still
> references VERSION.
> => Pass on VERSION via nmake explicitly to fix it being unset.

The recipe's tests didn't catch the issue due to Windows/CMD idiosyncrasies:
> The echo "..." retains the double quotes in its output.
> This in turn means that tclsh will treat it as a string and do the inner
> substitutions but not run the intended outer command.
> Which results in the test not exiting with a proper error code.

----
Unrelated, the builds for `osx-*` failed due to more stringent default error handling in newer Clang versions.
This can either be worked around by demoting them back to warnings or with a variant of the proposed patches here.
Feel free to adjust or replace the patches as you see fit!

----
Here's a checklist to do before merging.
- [x] Bump the build number if needed.

Fixes #12

conda-forge-webservices[bot] committed:
dummy commit for rerendering

Commit:
MNT: Re-rendered with conda-build 24.5.1, conda-smithy 3.36.2, and conda-forge-pinning 2024.06.21.08.07.40

mbargull committed:
Add 0002-Fix-incompatible-function-pointer-types-add-CONST86.patch

Signed-off-by: Marcel Bargull <marcel.bargull@udo.edu>

mbargull committed:
Add 0003-Fix-implicit-function-declaration-for-panic-on-macOS.patch

Signed-off-by: Marcel Bargull <marcel.bargull@udo.edu>

mbargull commented on recipe/0003-Fix-implicit-function-declaration-for-panic-on-macOS.patch:
```diff
@@ -0,0 +1,42 @@
+ 
+ #include "tkTable.h"
+ 
++#if defined(__APPLE__) && !defined(panic)
++/* On macOS, tcl.h does not #define panic Tcl_Panic. */
++# include <mach/mach.h>
++#endif
```
Alternatively (better?) we could replace the handful `panic` calls by `Tcl_Panic` -- that would make more sense to me.
The patch in its current state should represent the current behavior more, though.

mbargull commented on recipe/0002-Fix-incompatible-function-pointer-types-add-CONST86.patch:
```diff
@@ -0,0 +1,128 @@
+ #ifndef CONST84
+ #  define CONST84
+ #endif
++#ifndef CONST86
++# define CONST86 CONST84
++#endif
```
`Tk_CustomOption`'s member `Tk_OptionPrintProc` from Tk 8.6's `tk.h` has its return type defined as `CONST86 char *`.
This patch adds that `const`ness to the functions that are used as `Tk_OptionPrintProc` to avoid `incompatible-function-pointer-types` with default to errors in Clang>=16 or so.

mbargull committed:
Simplify test

Signed-off-by: Marcel Bargull <marcel.bargull@udo.edu>

mbargull committed:
Fix test on win

The echo "..." retains the double quotes in its output.
This in turn means that tclsh will treat it as a string and do the inner
substitutions but not run the intended outer command.
Which results in the test not exiting with a proper error code.

Signed-off-by: Marcel Bargull <marcel.bargull@udo.edu>

mbargull committed:
Fix VERSION/PACKAGE_VERSION usage for win builds

Tktable 2.8 had VERSION defined via generic/version.h which got then
reused in win/makefile.vc.
Tktable 2.10 has it renamed to PACKAGE_VERSION but win/makefile.vc still
references VERSION.
=> Pass on VERSION via nmake explicitly to fix it being unset.

Signed-off-by: Marcel Bargull <marcel.bargull@udo.edu>

mbargull committed:
Add stdlib("c")

Signed-off-by: Marcel Bargull <marcel.bargull@udo.edu>

mbargull commented:
@conda-forge-admin, please rerender

Commit:
MNT: Re-rendered with conda-build 24.5.1, conda-smithy 3.36.2, and conda-forge-pinning 2024.06.21.08.07.40

isuruf commented:
@conda-forge/tktable, can you review? This is one of the last missing things for r-base windows build.
